code/datums/components/scope.dm

```diff
@@ -0,0 +1,274 @@
+
+/datum/component/scope/proc/on_unwielded(obj/item/source, trait)
+	SIGNAL_HANDLER
+	var/mob/living/user = source.loc
+	stop_zooming(user)
+
+/datum/component/scope/proc/on_gun_fire(obj/item/gun/source, mob/living/user, atom/target, flag, params)
+	SIGNAL_HANDLER
+	if(!tracker?.given_turf || target == get_target(tracker.given_turf))
+		return NONE
+	INVOKE_ASYNC(source, TYPE_PROC_REF(/obj/item, afterattack), get_target(tracker.given_turf), user)
+	return COMPONENT_CANCEL_GUN_FIRE
+
+/datum/component/scope/proc/on_examine(datum/source, mob/user, list/examine_list)
+	SIGNAL_HANDLER
+
+	var/scope = istype(parent, /obj/item/gun) ? "scope in" : "zoom out"
+	switch(zoom_method)
+		if(ZOOM_METHOD_WIELD)
+			examine_list += "<span class='notice'>You can [scope] by wielding it with both hands.</span>"
+
+/**
+ * We find and return the best target to hit on a given turf.
+ *
+ * Arguments:
+ * * target_turf: The turf we are looking for targets on.
+*/
+/datum/component/scope/proc/get_target(turf/target_turf)
+	var/list/object_targets = list()
+	var/list/non_dense_targets = list()
+	for(var/atom/movable/possible_target in target_turf)
+		if(possible_target.layer <= PROJECTILE_HIT_THRESHHOLD_LAYER)
+			continue
+		if(possible_target.invisibility > tracker.owner.see_invisible)
+			continue
+		if(!possible_target.mouse_opacity)
+			continue
+		if(iseffect(possible_target))
+			continue
+		if(ismob(possible_target))
+			if(possible_target == tracker.owner)
+				continue
+			return possible_target
+		if(!possible_target.density)
+			non_dense_targets += possible_target
+			continue
+		object_targets += possible_target
+	for(var/obj/important_object as anything in object_targets)
+		return important_object
+	for(var/obj/unimportant_object as anything in non_dense_targets)
+		return unimportant_object
+	return target_turf
+
+/**
+ * We start zooming by adding our tracker overlay and starting our processing.
+ *
+ * Arguments:
+ * * user: The mob we are starting zooming on.
+*/
+/datum/component/scope/proc/zoom(mob/user)
+	if(isnull(user.client))
+		return
+	if(HAS_TRAIT(user, TRAIT_SCOPED))
+		to_chat(user, "<span class='warning'>You are already zoomed in!</span>")
+		return
+	if(attempting_to_scope)
+		to_chat(user, "<span class='warning'>You are already trying to scope in!</span>")
+		return
+	if(time_to_scope)
+		attempting_to_scope= TRUE
+		if(!do_after(user, time_to_scope, target = parent))
+			attempting_to_scope = FALSE
+			return
+		attempting_to_scope = FALSE
+	user.playsound_local(parent, 'sound/weapons/scope.ogg', 75, TRUE)
+	tracker = user.overlay_fullscreen("scope", /atom/movable/screen/fullscreen/cursor_catcher/scope, istype(parent, /obj/item/gun))
+	tracker.assign_to_mob(user, range_modifier)
+	if(allow_middle_click)
+		RegisterSignal(tracker, COMSIG_CLICK, PROC_REF(generic_click))
+	tracker_owner_ckey = user.ckey
+	if(user.is_holding(parent))
+		RegisterSignals(user, list(COMSIG_CARBON_SWAP_HANDS, COMSIG_PARENT_QDELETING), PROC_REF(stop_zooming))
```

Comment:
Haven't checked the code completely, but can you confirm that these are cleared properly all the time? Also when the gun gets deleted for instance

Reply:
I'll try deleting the gun, otherwise not seen any runtimes of them not being cleared or duplicated

Reply:
seems fine!